mixins/provider-mixin.md

```diff
@@ -0,0 +1,45 @@
+# ProviderMixin
+
+The `ProviderMixin` and `RequesterMixin` can be used to create a DI-like system within your app. Components can register themselves as providers of data (`ProviderMixin`), as well as request a provider of data via a key (`RequesterMixin`). Some more detail about this approach is provided in [this YouTube video](https://youtu.be/x9YDQUJx2uw?t=2104) from Google Developer Group's DevFest Nantes 2019.
+
+## Usage
+
+Apply the `ProviderMixin` to the component that will be responsible for providing some data to components that request it:
+
+```js
+import { ProviderMixin } from '@brightspace-ui/core/mixins/provider-mixin.js';
+
+class InterestingFactProvider extends ProviderMixin(LitElement) {
+	constructor() {
+		super();
+
+		this.provide('d2l-interesting-fact', 'Olives are not the same as fish');
```

Comment:
In practice, I expect this would more often provide some object eh?

Would we every expect that object to have dependencies of their own? And they could have dependencies?

It might be nice to have some concrete use-cases, perhaps for documentation, but also to sorta test this model.  Or maybe I am over thinking it?

Reply:
Yeah true, object would probably be a better example here. I'll change that.

I'm not sure if it's possible for dependencies to have dependencies, as I think that would imply the first-level dependency would itself be a LitElement object in order to implement the Requester mixin. Or, I guess that's possible, but unlikely? In the existing use cases [in `d2l-labs-attachment`](https://github.com/Brightspace/attachment/blob/e44cab1f0cecc55dd93acf59212fabc6872c0bd3/components/attachment.js#L99), this approach is used to return a cache object, a string, and a function - I can expand the demo to show each of these in action though, since that is definitely useful to see laid out!

Reply:
> 'm not sure if it's possible for dependencies to have dependencies, as I think that would imply the first-level dependency would itself be a LitElement object in order to implement the Requester mixin. 

Yeah, that's what I was contemplating as well.  If this were a DI system like the LMS, dependencies could certainly have dependencies, but this approach doesn't lend itself to that as-is.  And maybe that's ok though...? I don't want the doc to be more complex than needed... more just walking through concrete uses case to see how things hold up.

Reply:
I'm picturing this mostly being configured at the top-level of the app (in the app component itself probably). There, it can wire up a bunch of dependencies -- some of which might have other dependencies that it needs to pass in.

So not quite as amazing as a recursive DI system, but I think still pretty powerful given its simplicity.

Worst case, dependencies that depend on other things could be wired up as factories that the top-level app has already configured. Then the requesters just need to `.create()` the factories.

Reply:
Yeah, I think one of the benefits of the simplicity is that your `Provider` can provide anything, so long as it's able to be returned synchronously - if that means you have to call a `.create()` on it, or `await` on some part of whatever you get back as the `Requester`, that's not ideal but it's also not prevented by this approach.